### PR TITLE
Introduce `Iso8601Time` type to support `translation list` CLI command.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

#467 

# Overview

As part of supporting the `cardano-wallet transaction list` CLI command, we need to be able to accept optional `--before` and `--after` parameters, which specify a range of times. But in order to do that, we need to provide a way of parsing time values from textual input.

This PR:

- [x] adds a new type `Iso8601Time` which represents a point in time. The `FromText` instance parses a time in ISO 8601 format.
- [x] adds a round trip test to check that `Iso8601Time` values can be converted to and from `Text`.
- [x] adds tests to check that known-good ISO 8601 strings can be parsed successfully.
- [x] adds tests to check that known-bad ISO 8601 strings cannot be parsed successfully.
- [x] adds tests to ensure that equivalent ISO 8601 strings are parsed to values that are equivalent.

### CODE REVIEW

File: lib/cli/test/unit/Cardano/CLISpec.hs [hunk 1]
```diff
@@ -46,9 +57,18 @@ import qualified Data.Text.IO as TIO
...
         textRoundtrip $ Proxy @(Port "test")
         textRoundtrip $ Proxy @MnemonicSize
 
+    describe "Can decode basic ISO 8601 time examples" $ do
+        canDecodeIso8601Time "2008-09-15T15:53:00Z"
+        canDecodeIso8601Time "2008-09-15T15:53:00.1Z"
+        canDecodeIso8601Time "2008-09-15T15:53:00.12Z"
+        canDecodeIso8601Time "2008-09-15T15:53:00+05:00"
+        canDecodeIso8601Time "2008-09-15T15:53:00.1+05:00"
+        canDecodeIso8601Time "2008-09-15T15:53:00.12+05:00"
```
Comment (Anviking) [inline lib/cli/test/unit/Cardano/CLISpec.hs]: I believe the `-` makes this "extended ISO8601", which the **API** shouldn't support according to #466 (modulo I don't know why it says union signs):

> note: We do not accept and ISO-8601 extended format (with union signs) because it conflicts with the range separator

Regardless what the CLI does, maybe some more test-cases could be good
- `"20080915T15:53:00.1+05:00"` (edit: this was intended to be the basic format)
- `""`
?

Comment (KtorZ) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: > (modulo I don't know why it says union signs):

Probably some French jargon that doesn't translate well in English :grimacing: . I've changed _union_ for `-` to make it clearer.

File: lib/cli/cardano-wallet-cli.cabal [hunk 2]
```diff
@@ -41,11 +41,13 @@ library
     , fmt
     , http-client
     , iohk-monitoring
+    , iso8601-time
```
Comment (KtorZ) [inline lib/cli/cardano-wallet-cli.cabal]: I don't think this is a necessary dependency? 
What about:

https://hackage.haskell.org/package/time-1.9.2/docs/Data-Time-Format-ISO8601.html 

 ?

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 1 above)
Comment (KtorZ) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: Pointer: https://hackage.haskell.org/package/time-1.9.2/docs/Data-Time-Format-ISO8601.html#t:FormatExtension

Comment (Anviking) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: >  French jargon that doesn't translate well in English

I suspected that 😄

File: lib/cli/cardano-wallet-cli.cabal (hunk 2 above)
Comment (jonathanknowles) [reply lib/cli/cardano-wallet-cli.cabal]: > I don't think this is a necessary dependency?
> What about:
> 
> https://hackage.haskell.org/package/time-1.9.2/docs/Data-Time-Format-ISO8601.html

I also saw that. However, it depends on `>= time-1.9`, which currently isn't available on any version of Stackage. (The latest available is `1.8.0.2`.)

I'll experiment with using an `extra-dep` and report back.

Comment (jonathanknowles) [reply lib/cli/cardano-wallet-cli.cabal]: So it seems that many of the libraries on which we depend all require `time < 1.9`, even in the latest version of Stackage.

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: > Regardless what the CLI does, maybe some more test-cases could be good

Thanks for looking at this. I've added a bunch more test cases, including some negative test cases.

File: lib/cli/cardano-wallet-cli.cabal (hunk 2 above)
Comment (jonathanknowles) [reply lib/cli/cardano-wallet-cli.cabal]: It would be really nice to use this module (`Data.Time.Format.ISO8601`) from `time >= 1.9`, but I'm not sure it's worth the extra hassle until `time >= 1.9` is included in a Stackage LTS release.

I did (experimentally) try to get things to compile with `time-1.9.3` and succeeded in producing a build, but I had to resort to using `allow-newer: true` within `stack.yaml` to make this happen, which I'm sure we'd prefer to avoid.

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: > I believe the `-` makes this "extended ISO8601", which the **API** shouldn't support according to #466 (modulo I don't know why it says union signs):

That's an excellent point, one that I hadn't considered -- thanks for raising it.

We had a discussion about this in the slack channel. In the end, it was decided to allow the extended format in the CLI, as this format is easier to parse visually.

File: lib/cli/test/unit/Cardano/CLISpec.hs [hunk 3]
```diff
@@ -46,9 +57,54 @@ import qualified Data.Text.IO as TIO
...
+            canDecodeValidIso8601Time "2008-09-15T15:53:00.12+00:00"
+        describe "UTC+8 timezone (manually specified)" $ do
+            canDecodeValidIso8601Time "2008-09-15T15:53:00+08:00"
+            canDecodeValidIso8601Time "2008-09-15T15:53:00.1+08:00"
+            canDecodeValidIso8601Time "2008-09-15T15:53:00.12+08:00"
+        describe "UTC-8 timezone (manually specified)" $ do
+            canDecodeValidIso8601Time "2008-09-15T15:53:00-08:00"
+            canDecodeValidIso8601Time "2008-09-15T15:53:00.1-08:00"
+            canDecodeValidIso8601Time "2008-09-15T15:53:00.12-08:00"
+
```
Comment (Anviking) [inline lib/cli/test/unit/Cardano/CLISpec.hs]: It seems we don't support the basic format though. On slack you wrote "Or should we allow both Basic and Extended?" — is this intended?

If intended I hope people will not be confused by the API and CLI requiring different date-formats. I imagine a developer integrating with the API, that wants to pull values from a JSON log and try some requests with the CLI. Then the dates cannot simply be copy-pasted. But maybe that's a rare scenario, I have no idea.

Otherwise, I'd suggest at least one negative test to make it clear to the reader that we don't the support them:

Quick suggestion:

```suggestion

    describe "Cannot decode basic format ISO 8601 strings" $ do
        describe "UTC+0 timezone (Z)" $ do
            cannotDecodeInvalidIso8601Time "20190712T133821Z"
            cannotDecodeInvalidIso8601Time "20080915T1553001Z"
            cannotDecodeInvalidIso8601Time "20080915T15530012Z"
            
```

**Review by Anviking — APPROVED** [review]
LGTM! Seems we don't support the basic format, so a comment about that.

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 3 above)
Comment (KtorZ) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: > If intended I hope people will not be confused by the API and CLI requiring different date-formats. I imagine a developer integrating with the API, that wants to pull values from a JSON log and try some requests with the CLI. Then the dates cannot simply be copy-pasted.

I do agree. We should actually support the basic format in the CLI. We don't support the extended format in the API because it can be confusing, but it would be wrong to not support both in the CLI.

**Review by KtorZ — CHANGES_REQUESTED** [review]
It doesn't seem like we support decoding ISO-8601 date in the basic (non-extended) format?

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 3 above)
Comment (jonathanknowles) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: > I do agree. We should actually support the basic format in the CLI. We don't support the extended format in the API because it can be confusing, but it would be wrong to not support both in the CLI.

I agree that's a very sensible suggestion.

I've added support for ISO 8601 basic in addition to ISO 8601 extended. Because the `iso8601-time` package doesn't support ISO 8601 basic, I've removed that package and used the parsing functions from `Data.Time.Format` to add the necessary support. The `FromText` instance now decodes both basic and extended strings.

**Comment by jonathanknowles** [comment]
> It doesn't seem like we support decoding ISO-8601 date in the basic (non-extended) format?

I've now added support for ISO 8601 basic in addition to ISO 8601 extended. Because the `iso8601-time` package doesn't support ISO 8601 basic, I've removed that package and used the parsing functions from `Data.Time.Format` to add the necessary support. The `FromText` instance now decodes both basic and extended strings.

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 3 above)
Comment (jonathanknowles) [reply lib/cli/test/unit/Cardano/CLISpec.hs]: @Anviking wrote:
> If intended I hope people will not be confused by the API and CLI requiring different date-formats. I imagine a developer integrating with the API, that wants to pull values from a JSON log and try some requests with the CLI. Then the dates cannot simply be copy-pasted. But maybe that's a rare scenario, I have no idea.

I agree that we should support both. I've added support for the basic format, along with a set of tests to ensure that valid strings parse (and that various kinds of invalid strings do not). Thanks for highlighting this! :)

**Review by Anviking — APPROVED** [review]

**Review by KtorZ — APPROVED** [review]
